pyaerocom/colocateddata.py

```diff
@@ -81,55 +85,60 @@ class ColocatedData:
-                data = xarray.DataArray(data, **kwargs)
-                self.data = data
+    ###########################
+    ##   Pydantic ConfigDict
+    ###########################
+    model_config = ConfigDict(
+        arbitrary_types_allowed=True,
+        extra="allow",
+        protected_namespaces=(),
+        validate_assignment=True,
+    )
+
+    # versions lower than 1.0 are not based on Pydantic's BaseModel
+    __version__ = "1.0"
+
+    #################################
+    ##   Pydantic-based Attributes
+    #################################
+
+    data: Path | str | xr.DataArray | np.ndarray | None = None
+
+    @model_validator(mode="after")
+    def validate_data(self):
+        # if (
+        #     self.data is not None
+        # ):  # LB: Check that this is needed. may be covered by the model_validator
+        if isinstance(self.data, Path):
+            # make sure path is str instance
+            self.data = str(self.data)
+        if isinstance(self.data, str):
+            assert self.data.endswith("nc"), ValueError(
+                "Invalid data filepath str, must point to a .nc file"
+            )
+            self.open(self.data)
+        elif isinstance(self.data, xr.DataArray):
+            return self.data
+        elif isinstance(self.data, np.ndarray):
+            if self.data.ndim not in (3, 4):
+                raise DataDimensionError("invalid input, need 3D or 4D numpy array")
+            elif not self.data.shape[0] == 2:
+                raise DataDimensionError(
+                    "first dimension (data_source) must be of length 2(obs, model)"
+                )
+            if hasattr(self, "model_extra"):
+                da_keys = dir(xr.DataArray)
+                extra_args_from_class_initialization = {
+                    key: val for key, val in self.model_extra.items() if key in da_keys
+                }
             else:
-                raise ValueError(f"Failed to interpret input {data}")
-
-    @property
-    def data(self):
-        """:class:`xarray.DataArray` containing colocated data
-
-        Raises
-        ------
-        AttributeError
-            if data is not available
-
-        Returns
-        -------
-        xarray.DataArray
-            array containing colocated data and metadata (in fact, there is no
-            additional attributes to `ColocatedData` and everything is contained
-            in :attr:`data`).
-        """
-        if self._data is None:
-            raise AttributeError("No data available in this object")
-        return self._data
+                extra_args_from_class_initialization = {}
+            data = xr.DataArray(self.data, **extra_args_from_class_initialization)
+            self.data = data
 
-    @data.setter
-    def data(self, val):
-        if not isinstance(val, xarray.DataArray):
-            raise ValueError("Invalid input for data attribute, need instance of xarray.DataArray")
-        self._data = val
```

Comment:
I'd put the validation from the constructor in the setter, unless there is a good reason why validation shouldn't apply when using the setter directly.

Reply:
The validation is now handled by pydantic through type hinting, so this manual validation isn't needed anymore. Line 125 has the type hinting for what `data` can be, and if someone passes something else, it will crash.